### ISSUE TITLE
BUG: #18481 made BFGS minimization less accurate

### ISSUE THREAD

**Issue by jakevdp** [scipy/scipy#19024]
### Describe your issue.

#18481 appears to have made BFGS minimization less accurate. We noticed this in the JAX upstream nightly tests, which began failing after #18481 was merged: https://github.com/google/jax/actions/workflows/upstream-nightly.yaml

I suspect the dtype changes in `approx_derivative` are to blame.

### Reproducing Code Example

```python
import numpy as np
import scipy

def func(p):
    x, y = p
    return (x ** 2 + y - 11.) ** 2 + (x + y ** 2 - 7.) ** 2

x0 = np.ones(2, dtype='float32')
result = scipy.optimize.minimize(func, x0, method='BFGS')

print(f"{scipy.__version__=}")
print(result)
np.testing.assert_allclose(result.x, [3, 2])
```

### Error message

Output before #18481:
```
scipy.__version__='1.11.1'
  message: Optimization terminated successfully.
  success: True
   status: 0
      fun: 2.288511533827582e-15
        x: [ 3.000e+00  2.000e+00]
      nit: 8
      jac: [-1.907e-08  2.738e-08]
 hess_inv: [[ 1.612e-02 -9.549e-03]
            [-9.549e-03  3.517e-02]]
     nfev: 39
     njev: 13
```

Output after #18481:
```
scipy.__version__='1.12.0.dev0+1440.64be173'
  message: Desired error not necessarily achieved due to precision loss.
  success: False
   status: 2
      fun: 9.628441344453949e-06
        x: [ 3.000e+00  2.000e+00]
      nit: 6
      jac: [ 6.255e-04  2.977e-04]
 hess_inv: [[ 1.615e-02 -9.191e-03]
            [-9.191e-03  3.561e-02]]
     nfev: 291
     njev: 70
```
```pytb
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-1-d421a22eaf1d> in <cell line: 13>()
     11 result = scipy.optimize.minimize(func, x0, method='BFGS')
     12 print(result)
---> 13 np.testing.assert_allclose(result.x, [3, 2])

/usr/lib/python3.10/contextlib.py in inner(*args, **kwds)
     77         def inner(*args, **kwds):
     78             with self._recreate_cm():
---> 79                 return func(*args, **kwds)
     80         return inner
     81 

/usr/local/lib/python3.10/dist-packages/numpy/testing/_private/utils.py in assert_array_compare(comparison, x, y, err_msg, verbose, header, precision, equal_nan, equal_inf, strict)
    795                                 verbose=verbose, header=header,
    796                                 names=('x', 'y'), precision=precision)
--> 797             raise AssertionError(msg)
    798     except ValueError:
    799         import traceback

AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 2 / 2 (100%)
Max absolute difference: 0.00049761
Max relative difference: 0.00016587
 x: array([2.999502, 1.999956])
 y: array([3, 2])
```

### SciPy/NumPy/Python version and system information
The error is seen on the nightly scipy version: 
```shell
$ pip install -i https://pypi.anaconda.org/scientific-python-nightly-wheels/simple --pre -U scipy
```

**Comment by andyfaff** [comment]
@jakevdp, thank you for this report. I'm glad a jax(+scipy) developer is involved with this conversation, your expertise is vital here.

The PR was made in response to #15602 (important to read for background) that pointed out that `optimize` infrastructure was casting a lot of inputs to `np.float64` before evaluation. Thus, if the end user wished to work at a 32-bit precision this would be overridden by `optimize`, which would do all its work in 64-bit precision, and be sending float64 arrays back to the objective function for evaluation. The OP in #15602 mentioned that sometimes working in 32 bit precision was desired for various types of optimisation (e.g. deep learning frameworks).

The changes in #18481 are designed to respect the float width of the input array (or at least respect it more than previously). Specifically there are internal `ScalarFunction`, `VectorFunction` classes that take care of memoisation, grad/hess calculations. These functions were doing casting to float64, but now they remember the original dtype. Thus, if a 32-bit `x0` is passed to the original `minimize` call, then the input dtype is noted in `ScalarFunction`, and all future operations are done with that dtype. i.e. calls to the objective/grad/hess functions are done with float32. In this case (because callable(jac) is not provided) it means that you're doing numerical differentiation with 32-bit precision. This will be the cause of the *precision loss*.
If the input array is changed to 64-bit then the problem should not be present.

These changes to respect `x0.dtype` may not be consistent/perfect throughout the `minimize` stable, i.e. some minimizers may still cast to float-64. Whilst i haven't checked BFGS probably does nothing to change the bitness because it's pure Python.

In the original issue I did point out that:

> Most of them (i.e. gradient calculation) would require double precision to work properly.

the reply to that point was 

> all the deep learning frameworks use single precision by default or sometimes even lower in order to optimize the throughput.

What do you think about the changed behaviour @jakevdp? Will this have serious downstream harm?  Personally I do all my own work with float64, so won't experience issues. I'd like to hear about wider ramifications.

cc'ing @paulestano for contributions to conversation.

**Comment by andyfaff** [comment]
By judicious use of print debugging the logic flow the callstack is:

- `_optimize._minimize_bfgs` -> `_optimize._line_search_wolfe12` -> `_linesearch.line_search_wolfe1`, `_linesearch.scalar_search_wolfe1`. The last function calls out to [`minpack2.dcsrch`](https://github.com/scipy/scipy/blob/main/scipy/optimize/_linesearch.py#L155).

When the overall optimisation first starts `minpack2.dcsrch` returns `task` values of `b'CONVERGENCE'`, with finite values for `stp`. After several top level iterations `minpack2.dcsrch` returns a task of `b'WARNING: ROUNDING ERRORS PREVENT PROGRESS'`. At this point the linesearch is regarded to have failed, see [here](https://github.com/scipy/scipy/blob/main/scipy/optimize/_linesearch.py#L168) and [here](https://github.com/scipy/scipy/blob/main/scipy/optimize/_optimize.py#L1232), with `_optimize.line_search_wolfe12` returning `None`. BFGS tries to fallback to [`_linesearch.line_search_wolfe2`](https://github.com/scipy/scipy/blob/main/scipy/optimize/_optimize.py#L1240), but this fails as well. At this point the optimisation halts, and you get the top level minimisation halting.

Thus, the rounding errors seem to prevent BFGS working properly.

There are a few possible solutions:

- back out the changes made in #18481 
- intercept x0 in BFGS and make sure we use float64 internally in that minimize method
- try and change the linesearch code to work more easily with float32

At the current time I think BFGS and CG are going to be susceptible to this issue.

**Comment by jakevdp** [comment]
Thanks for the explanation! I think it makes sense to allow users to operate in f32, especially in a future world where the array API is supported and users can swap-in different backends in place of numpy. I can easily accommodate this downstream by casting to f64 or changing our test tolerances.

But I think as part of this change, it would be worth exploring whether the `scipy.opimize` machinery actually is functioning as expected in single-precision – this bug may be a sign that there are effectively (though probably not deliberately) some hard-coded assumptions about float precision within the implementations.

**Comment by andyfaff** [comment]
I'm guessing that judicious improvement of amin, xtol in [line_search_wolfe1 and scalar_search_wolfe1](https://github.com/scipy/scipy/blob/fe05d2e6c36c46d0afddba9f10f715979a5b7fd8/scipy/optimize/_linesearch.py#L33C59-L34C34) might improve the situation somewhat.

It looks like `amin` is approximately `np.sqrt(np.finfo(np.float64).eps)` (1.49e-8). The default value used in the linesearch code is 1e-8. 

`xtol` seems to be an order of magnitude larger than `np.finfo(np.float64).resolution`. The default value used in linesearch is 1e-14, but float64 resolution is 1e-15.

If my assumptions are correct it might be possible to adjust the line-search code to account for float precision. We need advice from someone who is familiar with linesearches though, this is not my area of expertise.

From what I can tell of the fortran code in [minpack2](https://github.com/scipy/scipy/tree/main/scipy/optimize/minpack2) there's nothing exploiting floating point precision in there. @ilayn, it looks like this minpack2 code should be relatively straightforward to port from Fortran to Python. There's only one goto.

**Comment by ilayn** [comment]
I'm almost done with PROPACK and looking not bad with ARPACK. Then I can take a stab at it. Looks like a relatively good and readable Fortran code.

**Comment by andyfaff** [comment]
I think the way forward for this is to:

1. respect user options (e.g. xrtol) if they provide them.
2. if jac is calculated via finite differences and `x0` is float32, then adjust `xtol`, `amax`, `amin` to sensible values for float32 as outlined above.
3. If jac is callable, then leave alone. The reason being that the OP problem seems to work when analytic derivatives are provided.

**Comment by paulestano** [comment]
Hello. 

I do believe this kind of behaviours to be expected: here you're combining finite difference and reduced precision arithmetic to estimate the gradients. Hence, you introduce 2 different noise sources in one of the most critical variables of the optimization process.

Adjusting `xtol` , `amax` and `amin` may mitigate some of the issues but I suspect there will still be some functions converging in $\texttt{FP64}$ and not $\texttt{FP32}$ using finite difference. This could also happen with poorly conditioned problems and/or with problems of bigger dimensions. In the paper I mentioned in #15602 the authors compute the gradients explicitly. If this is of interest it might be worth introducing some unit-tests with something like [CUTest](https://github.com/ralna/CUTEst).  

On another note, I noticed that this problem does converge when `finite_diff_rel_step` is manually set to `np.finfo(x0.dtype).eps` when $\texttt{FP32}$ is used and to  `np.finfo(x0.dtype).eps ** 0.5` when $\texttt{FP64}$ is used. I am not sure to fully understand this part of the code but it might be worth looking into [this](https://github.com/scipy/scipy/blob/fe05d2e6c36c46d0afddba9f10f715979a5b7fd8/scipy/optimize/_numdiff.py#L176) as a simple fix might just be to enforce  `finite_diff_rel_step=np.finfo(x0.dtype).eps if FP32 else np.finfo(x0.dtype).eps ** 0.5`.

Adapting previous code enforcing `finite_diff_rel_step=np.finfo(x0.dtype).eps` and the use of `2-point`.
```python
import numpy as np
import scipy


def func(p):
    x, y = p
    return (x**2 + y - 11.0) ** 2 + (x + y**2 - 7.0) ** 2

x0 = np.ones(2, dtype="float32")
type_info = np.finfo(x0.dtype) 
print("epsilon={}".format(type_info.eps))
jac = None
hess = None
result = scipy.optimize.minimize(
    func,
    x0,
    jac="2-point",
    hess=hess,
    method="BFGS",
    options={"finite_diff_rel_step": type_info.eps},
)

print(f"{scipy.__version__=}")
print(result)
np.testing.assert_allclose(result.x, [3, 2])
```

```
epsilon=1.1920928955078125e-07
scipy.__version__='1.12.0.dev0+1469.849c27d'
  message: Optimization terminated successfully.
  success: True
   status: 0
      fun: 2.9132250201047702e-12
        x: [ 3.000e+00  2.000e+00]
      nit: 11
      jac: [-7.470e-06 -3.397e-06]
 hess_inv: [[ 2.144e-02 -1.669e-02]
            [-1.669e-02  4.247e-02]]
     nfev: 71
     njev: 18
```

**Comment by andyfaff** [comment]
Thank @paulestano. I'm pretty sure that the optimal step sizes are being used for general finite differencing. The code you referred to in numdiff has been worked on a fair bit.

**Comment by paulestano** [comment]
Thank you for your answer @andyfaff.
I thought numdiff had only been used in $\texttt{FP64}$ before #18481 was merged. Was another part of scipy using numdiff
with $\texttt{FP32}$ inputs?

My argument was that, currently, if $\texttt{FP64}$ is used then [`_eps_for_method`](https://github.com/scipy/scipy/blob/fe05d2e6c36c46d0afddba9f10f715979a5b7fd8/scipy/optimize/_numdiff.py#L92) returns `EPS**0.5=1.49e-8` which is fine. However, in the case where $\texttt{FP32}$ is used, then `_eps_for_method` should return  `EPS**0.5=0.0003452` which I feel might be too big as a relative step. Once again, I'm not very familiar with the code and I might be missing something here.

**Comment by andyfaff** [comment]
`_numdiff.approx_derivative` has been respecting float width for a while, albeit with a few tidyups in 18481. The main change with that PR is that `ScalarFunction` (which uses `approx_derivative` for grad calculation) now also respects float width. `ScalarFunction` is used by many of the minimizer methods to deal with memoisation/fun/grad/hess calculations. Not all minimisation methods may fully respect float width, that's probably an ongoing project.

The step size calculation for numerical differentiation should be the same whatever the float width (i.e. balancing roundoff and secant error). i.e. for forward/reverse differences it's supposed to be sqrt(eps) (https://en.wikipedia.org/wiki/Numerical_differentiation).

I just tried adjusting amin, amax, xtol, in the manner I described above. There are still roundoff errors when using the minimiser. I find it interesting that L-BFGS-B doesn't have this issue.

What's happening is that further improvement of the step [isn't found](https://github.com/scipy/scipy/blob/5ebc7afe8949d097728ee5b93271a481f2fcb881/scipy/optimize/minpack2/dcsrch.f#L319), which [triggers the warning](https://github.com/scipy/scipy/blob/main/scipy/optimize/minpack2/dcsrch.f#L240). If one looks at the example given in the first post the solution is actually pretty good. It's just classified as a fail because of the warning message being raised.

**Comment by paulestano** [comment]
Hello @andyfaff. Thank you for the clarifications on numerical differentiation.
I also noted that L-BFGS-B does converge with no warning. After a little bit of debugging it seems like L-BFGS-B still does not respect float width (or at least typing). I manually logged the types of [`self.x` and `self.x_dtype` in ScalarFunction class](https://github.com/scipy/scipy/blob/8b473b950fde84f06201ff610cadecd5859f019b/scipy/optimize/_differentiable_functions.py#L115)

Running the snippet from above using L-BFGS-B I get:
```
self.x_dtype=dtype('float64')
self.x.dtype=dtype('float64')
```
While for BFGS 
```
self.x_dtype=dtype('float32')
self.x.dtype=dtype('float32')
```
You can also see that the output of the snippet from above is actually the same in single and double precision.
This probably comes from the fact that L-BFGS-B fortran code [does expect double precision inputs](https://github.com/scipy/scipy/blob/5ebc7afe8949d097728ee5b93271a481f2fcb881/scipy/optimize/lbfgsb_src/lbfgsb.f#L49)  and it could also potentially explain the better stability of L-BFGS-B. 

I am also looking at potential fine-tuned values for `xtol` , `amax` and `amin` but I believe that modifying these values might generate new warnings within the fortran code. For instance,   if we increase `xtol` to account for FP32 precision we may hit $x \leq xtol$ and this would also [trigger a warning (not a roundoff one though)](https://github.com/scipy/scipy/blob/5ebc7afe8949d097728ee5b93271a481f2fcb881/scipy/optimize/minpack2/dcsrch.f#L241). Same goes if we hit `amin` or `amax` while the wolfe conditions are not reached. This could potentially explain the former very low (respectively high) values for `amin` (resp. `amax`).

**Comment by dschmitz89** [comment]
The limitations of finite difference gradients are well known, I am not surprised that it can alead to optimization failures for simple functions in single precision. Aligning the possible types of gradient/objective function values across all of `scipy.optimize` is not possible due to the difficulties with the old Fortran libraries unfortunately.

My take on this: let's add a warning in the documentation that if no analytical gradient can be provided, the approximated gradient might become inaccurate.

**Comment by paulestano** [comment]
@dschmitz89 I think the limitations of finite difference are well known but I'm interested in the interactions between the approximation errors (from numerical differentiation) with the arithmetic errors (coming from finite precision computation). It seems like for double precision arithmetic errors are neglectible when using numerical differentiation while it's not the case when using single precision.
Hence, I do believe we should add a warning but only when FP32 is used with finite difference, as it seems to be working fine in FP64.

**Comment by dschmitz89** [comment]
@paulestano : sorry for the late reply. By warning, do you mean throwing a warning at runtime at the beginning of the optimization if the gradient is approximated in single precision? That is an option but could also be perceived as spam by users sd often single precision is good enough. Any strong opinion here @andyfaff ?

**Comment by andyfaff** [comment]
It's a messy situation. I assume that respecting float width is important for some users.

- I'm pretty sure that the finite difference code is choosing sensible step sizes for both float widths
- However, the range of minimisers in `optimize` were not originally coded with adaptability to float width in mind.
- Some codes may be more easily adapted to deal with float width than others (i.e. Fortran based minimisers).
-For BFGS (mostly Python) I don't think the algorithm is natively hostile to single precision. I *guess* that adapting `c1`/`c2` may yield some relief when working in single precision. But I don't know how to choose a better `c1`/`c2` for single precision compared to double precision.

I think a sensible route is amend documentation to point out pitfalls first. Then figure out better c1/c2/amin/amax values. If it still causes issues we can think about a warning.

**Comment by andyfaff** [comment]
I had a debugging session on this issue this morning, with the OP problem. Adjusting `c1`/`c2` didn't solve the issue.

I then started playing around with the same problem in double precision to try to get it to fail. You can achieve the same 'failure' (i.e. precision loss) by setting `gtol` to a low value, e.g. 1e-11. If this `gtol` value is set too low then the BFGS code continues to do a line search (https://github.com/scipy/scipy/blob/main/scipy/optimize/_optimize.py#L1488), but in a region where the dsrch code will fail because of precision losses.

I therefore think the ultimate cause is that the default `gtol` is set too low for this particular problem in single precision; setting `gtol=1e-3` allows the single precision to complete without issue.

Thus, I think the solution to this issue is to amend the documentation to say something along the lines of:

- if you see `Desired error not necessarily achieved due to precision loss`, then consider setting `gtol` to a higher value. The precision losses typically arise if the finite differences differentiation cannot provide sufficient precision to satisfy the `gtol` termination criterion; typically when working in single precision (i.e. a callable jac is not provided).

**Comment by paulestano** [comment]
I think adjusting `c1` and/or `c2` to a low precision setting might be interesting but I can't find any relevant publication on it.
I think `gtol=1e-5` might be a bit low indeed when you combine single precision and finite differences.
As a comparison `sklearn` uses a tolerance of [`tol=1e-4` in double precision for logistic regression problems (convex problem)](https://github.com/scikit-learn/scikit-learn/blob/3b35ad03d348658cd85ba4b3647873fae5f75a48/sklearn/linear_model/_logistic.py#L464).

**Comment by andyfaff** [comment]
closed by #19259
